File: ingest_train/stockvision/features.py
"""Technical indicators.

The C++ server rebuilds these same 17 columns in the same order, so a model
trained here can be served by either backend.
"""

import logging

import numpy as np
import pandas as pd

from . import config

log = logging.getLogger(__name__)


def simple_moving_average(series, window):
    return series.rolling(window=window, min_periods=1).mean()


def exponential_moving_average(series, span):
    return series.ewm(span=span, adjust=False).mean()


def relative_strength_index(series, period=14):
    change = series.diff()
    gain = change.clip(lower=0.0)
    loss = -change.clip(upper=0.0)
    average_gain = gain.ewm(alpha=1 / period, adjust=False).mean()
    average_loss = loss.ewm(alpha=1 / period, adjust=False).mean()
    strength = average_gain / average_loss.replace(0, np.nan)
    # A flat series has no losses, which leaves NaN; 50 means "neither overbought nor oversold".
    return (100 - 100 / (1 + strength)).fillna(50.0)


def macd(series, fast=12, slow=26, signal=9):
    line = exponential_moving_average(series, fast) - exponential_moving_average(series, slow)
    signal_line = exponential_moving_average(line, signal)
    return pd.DataFrame(
        {"macd": line, "macd_signal": signal_line, "macd_hist": line - signal_line}
    )


def bollinger_bands(series, window=20, num_std=2.0):
    middle = simple_moving_average(series, window)
    spread = num_std * series.rolling(window=window, min_periods=1).std().fillna(0.0)
    return pd.DataFrame(
        {"bb_upper": middle + spread, "bb_lower": middle - spread, "bb_width": 2 * spread}
    )


def log_returns(series):
    return np.log(series / series.shift(1)).fillna(0.0)


def build_feature_frame(df, ticker=None):
    """Turn raw OHLCV bars into the feature columns the models are trained on."""
    if "Close" not in df.columns:
        raise KeyError("DataFrame must contain a 'Close' column")

    frame = pd.DataFrame(index=df.index)
    frame["close"] = df["Close"].astype(float)
    frame["open"] = df["Open"].astype(float)
    frame["high"] = df["High"].astype(float)
    frame["low"] = df["Low"].astype(float)
    frame["volume"] = df["Volume"].astype(float)
    frame["log_return"] = log_returns(frame["close"])
    frame["sma_10"] = simple_moving_average(frame["close"], 10)
    frame["sma_30"] = simple_moving_average(frame["close"], 30)
    frame["ema_12"] = exponential_moving_average(frame["close"], 12)
    frame["ema_26"] = exponential_moving_average(frame["close"], 26)
    frame["rsi_14"] = relative_strength_index(frame["close"], 14)
    frame = pd.concat([frame, macd(frame["close"]), bollinger_bands(frame["close"])], axis=1)
    frame = frame.bfill().ffill().fillna(0.0)

    if config.USE_FEAST and ticker:
        frame = _overlay_feast_row(frame, ticker)
    return frame


# Feast stores the same indicators under its own names.
_FEAST_COLUMNS = {
    "RSI_14": "rsi_14",
    "MACD_12_26": "macd",
    "BB_upper_20": "bb_upper",
    "BB_lower_20": "bb_lower",
    "SMA_5": "sma_10",
    "SMA_20": "sma_30",
}


def _overlay_feast_row(frame, ticker):
    """Replace the newest indicator row with the one served by the Feast online store."""
    try:
        from feature_store.feature_client import get_features

        row = get_features(ticker).iloc[0]
    except Exception as exc:
        log.warning("Feast lookup failed for %s, using inline features: %s", ticker, exc)
        return frame

    for feast_name, column in _FEAST_COLUMNS.items():
        if feast_name in row:
            frame.loc[frame.index[-1], column] = float(row[feast_name])
    return frame

File: cpp_server/src/features.cpp
#include "features.h"

#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace stockvision {
namespace {

using Series = std::vector<double>;

Series moving_average(const Series& values, int window) {
    Series out(values.size());
    double running_total = 0.0;
    for (std::size_t i = 0; i < values.size(); ++i) {
        running_total += values[i];
        if (static_cast<int>(i) >= window) running_total -= values[i - window];
        out[i] = running_total / std::min(window, static_cast<int>(i) + 1);
    }
    return out;
}

Series exponential_average(const Series& values, int span) {
    Series out(values.size());
    if (values.empty()) return out;
    const double weight = 2.0 / (span + 1.0);
    out[0] = values[0];
    for (std::size_t i = 1; i < values.size(); ++i) {
        out[i] = weight * values[i] + (1.0 - weight) * out[i - 1];
    }
    return out;
}

Series relative_strength_index(const Series& close, int period) {
    Series out(close.size(), 50.0);  // 50 is neutral, used until we have two bars
    const double weight = 1.0 / period;
    double average_gain = 0.0;
    double average_loss = 0.0;

    for (std::size_t i = 1; i < close.size(); ++i) {
        const double change = close[i] - close[i - 1];
        const double gain = std::max(change, 0.0);
        const double loss = std::max(-change, 0.0);
        // The first bar seeds the averages; after that they decay exponentially.
        average_gain = i == 1 ? gain : weight * gain + (1.0 - weight) * average_gain;
        average_loss = i == 1 ? loss : weight * loss + (1.0 - weight) * average_loss;
        out[i] = average_loss <= 1e-12 ? 100.0
                                       : 100.0 - 100.0 / (1.0 + average_gain / average_loss);
    }
    return out;
}

Series rolling_std(const Series& values, int window) {
    Series out(values.size());
    for (std::size_t i = 0; i < values.size(); ++i) {
        const std::size_t first = i + 1 > static_cast<std::size_t>(window) ? i + 1 - window : 0;
        const std::size_t count = i + 1 - first;
        if (count <= 1) continue;  // a single point has no spread

        double mean = 0.0;
        for (std::size_t j = first; j <= i; ++j) mean += values[j];
        mean /= count;

        double variance = 0.0;
        for (std::size_t j = first; j <= i; ++j) variance += (values[j] - mean) * (values[j] - mean);
        out[i] = std::sqrt(variance / (count - 1));
    }
    return out;
}

}  // namespace

FeatureMatrix build_features(const PriceTable& prices) {
    const std::size_t days = prices.size();
    if (days == 0) throw std::runtime_error("No price rows to build features from");

    const Series& close = prices.close;

    Series log_return(days);
    for (std::size_t i = 1; i < days; ++i) {
        if (close[i - 1] > 0.0) log_return[i] = std::log(close[i] / close[i - 1]);
    }

    const Series sma_10 = moving_average(close, 10);
    const Series sma_30 = moving_average(close, 30);
    const Series ema_12 = exponential_average(close, 12);
    const Series ema_26 = exponential_average(close, 26);
    const Series rsi_14 = relative_strength_index(close, 14);

    Series macd(days);
    for (std::size_t i = 0; i < days; ++i) macd[i] = ema_12[i] - ema_26[i];
    const Series macd_signal = exponential_average(macd, 9);

    const Series middle_band = moving_average(close, 20);
    const Series spread = rolling_std(close, 20);

    // The order here has to match FEATURE_NAMES in params.txt.
    FeatureMatrix matrix;
    matrix.names = {"close", "open", "high", "low", "volume", "log_return",
                    "sma_10", "sma_30", "ema_12", "ema_26", "rsi_14",
                    "macd", "macd_signal", "macd_hist",
                    "bb_upper", "bb_lower", "bb_width"};
    matrix.rows.resize(days);

    for (std::size_t i = 0; i < days; ++i) {
        const double bb_upper = middle_band[i] + 2.0 * spread[i];
        const double bb_lower = middle_band[i] - 2.0 * spread[i];
        matrix.rows[i] = {
            static_cast<float>(close[i]),
            static_cast<float>(prices.open[i]),
            static_cast<float>(prices.high[i]),
            static_cast<float>(prices.low[i]),
            static_cast<float>(prices.volume[i]),
            static_cast<float>(log_return[i]),
            static_cast<float>(sma_10[i]),
            static_cast<float>(sma_30[i]),
            static_cast<float>(ema_12[i]),
            static_cast<float>(ema_26[i]),
            static_cast<float>(rsi_14[i]),
            static_cast<float>(macd[i]),
            static_cast<float>(macd_signal[i]),
            static_cast<float>(macd[i] - macd_signal[i]),
            static_cast<float>(bb_upper),
            static_cast<float>(bb_lower),
            static_cast<float>(bb_upper - bb_lower),
        };
    }
    return matrix;
}

}  // namespace stockvision
